internal/graph/membershipset.go

```diff
@@ -0,0 +1,254 @@
+// from both the parent relationship and the resource's result itself, assuming either have a caveat
+// associated.
+func (ms *MembershipSet) AddMemberViaRelationship(
+	resourceID string,
+	resourceCaveatExpression *v1.CaveatExpression,
+	parentRelationship *core.RelationTuple,
+) {
+	intersection := caveatAnd(wrapCaveat(parentRelationship.Caveat), resourceCaveatExpression)
+	ms.addMember(resourceID, intersection)
+}
+
+func (ms *MembershipSet) addMember(resourceID string, caveatExpr *v1.CaveatExpression) {
+	existing, ok := ms.membersByID[resourceID]
+	if !ok {
+		ms.hasDeterminedMember = ms.hasDeterminedMember || caveatExpr == nil
+		ms.membersByID[resourceID] = caveatExpr
+		return
+	}
+
+	// If a determined membership result has already been found (i.e. there is no caveat),
+	// then nothing more to do.
+	if existing == nil {
+		return
+	}
+
+	// If the new caveat expression is nil, then we are adding a determined result.
+	if caveatExpr == nil {
+		ms.hasDeterminedMember = true
+		ms.membersByID[resourceID] = nil
+		return
+	}
+
+	// Otherwise, the caveats get unioned together.
+	ms.membersByID[resourceID] = caveatOr(existing, caveatExpr)
+}
+
+// UnionWith combines the results found in the given map with the members of this set.
+// The changes are made in-place.
+func (ms *MembershipSet) UnionWith(resultsMap CheckResultsMap) {
+	for resourceID, details := range resultsMap {
+		ms.addMember(resourceID, details.Expression)
+	}
+}
+
+// IntersectWith intersects the results found in the given map with the members of this set.
+// The changes are made in-place.
+func (ms *MembershipSet) IntersectWith(resultsMap CheckResultsMap) {
+	for resourceID, details := range resultsMap {
+		existing, ok := ms.membersByID[resourceID]
+		if !ok {
+			continue
+		}
+
+		ms.membersByID[resourceID] = caveatAnd(existing, details.Expression)
+	}
+
+	hasDeterminedMember := false
+	for resourceID, expr := range ms.membersByID {
+		if _, ok := resultsMap[resourceID]; !ok {
+			delete(ms.membersByID, resourceID)
+			continue
+		}
+
+		if expr == nil {
+			hasDeterminedMember = true
+		}
+	}
+
+	ms.hasDeterminedMember = hasDeterminedMember
+}
+
+// Subtract subtracts the results found in the given map with the members of this set.
+// The changes are made in-place.
+func (ms *MembershipSet) Subtract(resultsMap CheckResultsMap) {
+	for resourceID, details := range resultsMap {
+		existing, ok := ms.membersByID[resourceID]
+		if !ok {
+			continue
+		}
+
+		// If the incoming result has no caveat, then this removal is absolute.
+		if details.Expression == nil {
+			delete(ms.membersByID, resourceID)
+			continue
+		}
+
+		// Otherwise, the caveat expression gets combined with an intersection of the inversion
+		// of the expression.
+		ms.membersByID[resourceID] = caveatSub(existing, details.Expression)
+	}
+
+	ms.hasDeterminedMember = false
+	for _, expr := range ms.membersByID {
+		if expr == nil {
+			ms.hasDeterminedMember = true
+			break
+		}
+	}
```

Comment:
We can avoid doing two loops i we iterate on the `membersByID` first:

```go
	ms.hasDeterminedMember = false
	for resourceID, expression := range ms.membersByID {
		if details, ok := resultsMap[resourceID]; ok {
			// If the incoming member has no caveat, then this removal is absolute.
			if details.Expression == nil {
				delete(ms.membersByID, resourceID)
				continue
			}

			// Otherwise, the caveat expression gets combined with an intersection of the inversion
			// of the expression.
			ms.membersByID[resourceID] = caveatSub(expression, details.Expression)
		} else {
			if expression == nil {
				ms.hasDeterminedMember = true
			}
		}
	}
```